File: packages/mobile-android/sdk/src/main/java/com/rtcexpress/sdk/SignalingClient.kt
package com.rtcexpress.sdk

import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject

class SignalingClient(
    private val http: OkHttpClient = OkHttpClient()
) {
    interface Listener {
        fun onConnected(userId: String)
        fun onDisconnected()
        fun onRoomJoined(roomId: String, members: List<String>)
        fun onUserJoined(roomId: String, userId: String)
        fun onUserLeft(roomId: String, userId: String)
        fun onMessage(message: RoomMessage)
        fun onCallInvite(invite: CallInvite)
        fun onCallAccept(payload: JSONObject)
        fun onCallReject(payload: JSONObject)
        fun onCallEnd(payload: JSONObject)
        fun onWebRtcOffer(payload: JSONObject)
        fun onWebRtcAnswer(payload: JSONObject)
        fun onIceCandidate(payload: JSONObject)
        fun onError(message: String)
    }

    private var socket: WebSocket? = null
    var listener: Listener? = null

    fun connect(serverUrl: String, token: String) {
        val wsBase = serverUrl.trimEnd('/').replace("^http".toRegex(), "ws")
        val request = Request.Builder()
            .url("$wsBase/ws?token=${java.net.URLEncoder.encode(token, Charsets.UTF_8.name())}")
            .build()
        socket = http.newWebSocket(request, object : WebSocketListener() {
            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(JSONObject(text))
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                listener?.onDisconnected()
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                listener?.onError(t.message ?: "WebSocket failed")
                listener?.onDisconnected()
            }
        })
    }

    fun send(type: String, payload: JSONObject) {
        val ws = socket ?: throw IllegalStateException("Not connected")
        ws.send(JsonUtil.clientMessage(type, payload))
    }

    fun close() {
        socket?.close(1000, "bye")
        socket = null
    }

    private fun handleMessage(message: JSONObject) {
        val type = message.getString("type")
        val payload = message.optJSONObject("payload") ?: JSONObject()
        when (type) {
            "connected" -> listener?.onConnected(payload.optString("userId"))
            "room_joined" -> {
                val members = mutableListOf<String>()
                val arr = payload.optJSONArray("members")
                if (arr != null) for (i in 0 until arr.length()) members.add(arr.getString(i))
                listener?.onRoomJoined(payload.getString("roomId"), members)
            }
            "user_joined" -> listener?.onUserJoined(payload.getString("roomId"), payload.getString("userId"))
            "user_left" -> listener?.onUserLeft(payload.getString("roomId"), payload.getString("userId"))
            "message" -> listener?.onMessage(
                RoomMessage(
                    roomId = payload.getString("roomId"),
                    fromUserId = payload.getString("fromUserId"),
                    text = payload.getString("text"),
                    sentAt = payload.optLong("sentAt")
                )
            )
            "call_invite" -> listener?.onCallInvite(
                CallInvite(
                    callId = payload.getString("callId"),
                    roomId = payload.getString("roomId"),
                    fromUserId = payload.getString("fromUserId"),
                    toUserId = payload.getString("toUserId"),
                    callType = payload.optString("callType", "voice")
                )
            )
            "call_accept" -> listener?.onCallAccept(payload)
            "call_reject" -> listener?.onCallReject(payload)
            "call_end" -> listener?.onCallEnd(payload)
            "webrtc_offer" -> listener?.onWebRtcOffer(payload)
            "webrtc_answer" -> listener?.onWebRtcAnswer(payload)
            "ice_candidate" -> listener?.onIceCandidate(payload)
            "error" -> listener?.onError(payload.optString("message", "Unknown error"))
        }
    }
}

File: packages/mobile-ios/Sources/RTCExpress/SignalingClient.swift
import Foundation

protocol SignalingListener: AnyObject {
    func onConnected(userId: String)
    func onDisconnected()
    func onRoomJoined(roomId: String, members: [String])
    func onMessage(_ message: RoomMessage)
    func onCallInvite(_ invite: CallInvite)
    func onCallAccept(_ payload: [String: Any])
    func onCallReject(_ payload: [String: Any])
    func onCallEnd(_ payload: [String: Any])
    func onWebRtcOffer(_ payload: [String: Any])
    func onWebRtcAnswer(_ payload: [String: Any])
    func onIceCandidate(_ payload: [String: Any])
    func onError(_ message: String)
}

final class SignalingClient {
    weak var listener: SignalingListener?
    private var task: URLSessionWebSocketTask?
    private let session = URLSession(configuration: .default)

    func connect(serverUrl: String, token: String) {
        let wsBase = serverUrl
            .trimmingCharacters(in: .init(charactersIn: "/"))
            .replacingOccurrences(of: "https://", with: "wss://")
            .replacingOccurrences(of: "http://", with: "ws://")
        let encoded = token.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? token
        let url = URL(string: "\(wsBase)/ws?token=\(encoded)")!
        task = session.webSocketTask(with: url)
        task?.resume()
        listen()
    }

    func send(type: String, payload: [String: Any]) throws {
        guard let task else { throw NSError(domain: "RTCExpress", code: 2) }
        let text = try JsonCodec.clientMessage(type: type, payload: payload)
        task.send(.string(text)) { _ in }
    }

    func close() {
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
    }

    private func listen() {
        task?.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .failure:
                self.listener?.onDisconnected()
            case .success(let message):
                if case .string(let text) = message,
                   let data = text.data(using: .utf8),
                   let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    self.handle(json)
                }
                self.listen()
            }
        }
    }

    private func handle(_ message: [String: Any]) {
        guard let type = message["type"] as? String else { return }
        let payload = message["payload"] as? [String: Any] ?? [:]
        switch type {
        case "connected":
            listener?.onConnected(userId: payload["userId"] as? String ?? "")
        case "room_joined":
            listener?.onRoomJoined(
                roomId: payload["roomId"] as? String ?? "",
                members: payload["members"] as? [String] ?? []
            )
        case "message":
            listener?.onMessage(RoomMessage(
                roomId: payload["roomId"] as? String ?? "",
                fromUserId: payload["fromUserId"] as? String ?? "",
                text: payload["text"] as? String ?? "",
                sentAt: payload["sentAt"] as? Int64 ?? 0
            ))
        case "call_invite":
            listener?.onCallInvite(CallInvite(
                callId: payload["callId"] as? String ?? "",
                roomId: payload["roomId"] as? String ?? "",
                fromUserId: payload["fromUserId"] as? String ?? "",
                toUserId: payload["toUserId"] as? String ?? "",
                callType: payload["callType"] as? String ?? "voice"
            ))
        case "call_accept": listener?.onCallAccept(payload)
        case "call_reject": listener?.onCallReject(payload)
        case "call_end": listener?.onCallEnd(payload)
        case "webrtc_offer": listener?.onWebRtcOffer(payload)
        case "webrtc_answer": listener?.onWebRtcAnswer(payload)
        case "ice_candidate": listener?.onIceCandidate(payload)
        case "error":
            listener?.onError(payload["message"] as? String ?? "Unknown error")
        default: break
        }
    }
}
